### PR TITLE
[dynamo] Ensure Dynamo uses this graph's fakes for `Tensor` `example_value`s

### LINKED ISSUE #111869: [dynamo] `get_fake_value` doesn't always return fake values
### 🐛 Describe the bug

In particular, when you run `get_fake_value` on a `getattr_var`, it will return the real tensor. This has got to be wrong. If only because the function is named `get_fake_value`.

Related: https://github.com/pytorch/pytorch/pull/111864

### Versions

main

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @eellison @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### LINKED ISSUE #111873: [dynamo] Should `wrap_to_fake_tensor_and_record` be idempotent with regard to `FakeTensor`s?
### 🐛 Describe the bug

`wrap_to_fake_tensor_and_record` will actually wrap an existing `FakeTensor` within another layer of `FakeTensor`. What use does this have?

Shouldn't this function be idempotent - return the same `FakeTensor`, and not wrap it further? Further, shouldn't we forbid wrapping one fake tensor in another? 

Perhaps there are use cases, since one layer of fake tensor may have metadata that another would not?

Nevertheless, for the purpose of Dynamo, perhaps we should not allow nested fake tensor.

This is because nested FakeTensor could lead to weird aliasing bugs, as we are now shifting to using `FakeTensor` object identity to identify aliasing when it comes to user code in Dynamo. (See: https://github.com/pytorch/pytorch/issues/111585 https://github.com/pytorch/pytorch/pull/111196)

### Versions

main

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @eellison @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/111869, Fixes (detailed list of cases handled): https://github.com/pytorch/pytorch/pull/111913#discussion_r1370267313, fully fixes: https://github.com/pytorch/pytorch/issues/111873

Adds sanity checks ensuring that Dynamo uses this graph's fakes for Tensor `example_values`

Handles the main (and only?) entrypoints for new `FakeTensor`s in a Dynamo graph:
- `wrap_fx_proxy_cls`
- `VariableTracker.wrap_tensor` 

Ensures that `get_fake_value` returns a fake except when we know we are going to properly wrap non-fakes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -1364,17 +1364,24 @@ def get_fake_value(node, tx):
     if "example_value" in node.meta and is_fake(node.meta["example_value"]):
         return node.meta["example_value"]
 
-    def fake_wrapper(e):
+    def ensure_fake(e):
         if isinstance(e, torch.Tensor):
             assert is_fake(e)
```
Comment (jon-chuang) [inline torch/_dynamo/utils.py:1369]: Should we also ensure that the fake of the args is this graph's fake?

Comment (jon-chuang) [reply torch/_dynamo/utils.py:1369]: I think the answer is yes - as long as we can get away with it, stricter is better unless someone has a good reason to make an exception.

Comment (jon-chuang) [reply torch/_dynamo/utils.py]: Let's monitor if anything goes wrong as a result

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -1117,10 +1117,9 @@ def wrap_tensor(self, value: torch.Tensor):
         # TODO: I think the result is guaranteed to be fake with
         # ignore_subclass changes
         # Note: this information is conveyed via subclass_type now
-        fake_tensor_value = None
-        example_value = tensor_variable.proxy.node.meta["example_value"]
-        if is_fake(example_value):
-            fake_tensor_value = example_value
+        fake_tensor_value = tensor_variable.proxy.node.meta["example_value"]
+        if not is_fake(fake_tensor_value):
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builder.py:1121]: Shouldn't this be stronger - this graph's fakes?

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -209,6 +209,10 @@ def maybe_get_fake_mode(t):
         m = modes[0]
         assert all(m is x for x in modes)
         return m
+    elif isinstance(t, torch.Tensor) and torch._is_functional_tensor(t):
+        reapply_views = torch._C._functionalization_reapply_views_tls()
+        unwrapped = torch._C._functorch._unwrap_functional_tensor(t, reapply_views)
+        return maybe_get_fake_mode(unwrapped)
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:215]: oh?! What did you need this for?

File: torch/_dynamo/variables/builder.py [hunk 4]
```diff
@@ -1390,22 +1393,16 @@ def wrap_fx_proxy_cls(
...
 
-    def _is_functional_tensor_fakified_by_dynamo(x):
-        if isinstance(x, torch.Tensor) and torch._is_functional_tensor(x):
-            reapply_views = torch._C._functionalization_reapply_views_tls()
-            unwrapped = torch._C._functorch._unwrap_functional_tensor(x, reapply_views)
-            return (
-                isinstance(unwrapped, FakeTensor)
-                and unwrapped.fake_mode == tx.fake_mode
-            )
-        return False
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:1401]: moved from here

**Review by ezyang — APPROVED** [review]
lgtm if it passes

File: torch/_subclasses/fake_tensor.py (hunk 3 above)
Comment (jon-chuang) [reply torch/_subclasses/fake_tensor.py:215]: As you can see, it merely completes `maybe_get_fake_mode` to be on par with `is_fake` when the tensor is functional.

File: torch/_dynamo/variables/builder.py [hunk 5]
```diff
@@ -1415,13 +1412,12 @@ def _clone_input(value):
 
     with preserve_rng_state():
         if example_value is None:
-            example_value = get_fake_value(proxy.node, tx)
+            # only allow_non_graph_fake in this instance because we handle the non-fake
+            # cases properly below.
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply torch/_dynamo/variables/builder.py:1416]: When we call `var_getattr` on an nn.Module. We then want to properly identify the fake tensor's source to be the nn.Module.

**Pull request closed without merging**